File: dictatorscript-vscode/server/src/lexer.ts
/**
 * DictatorScript Lexer
 *
 * TypeScript port of src/compiler/Lexer.cpp.
 * Tokenizes DictatorScript source into a stream of tokens.
 * Positions are 0-based (LSP convention).
 */

import { TokenType, KEYWORD_MAP } from '../../shared/src/tokenType';
import { Token } from '../../shared/src/token';
import { LexerDiagnostic } from '../../shared/src/types';
import { VALID_ESCAPE_CHARS } from '../../shared/src/languageData';

export interface LexResult {
    tokens: Token[];
    diagnostics: LexerDiagnostic[];
}

export function tokenize(source: string): LexResult {
    const lexer = new Lexer(source);
    return lexer.tokenize();
}

class Lexer {
    private source: string;
    private pos = 0;
    private line = 0;
    private col = 0;
    private tokens: Token[] = [];
    private diagnostics: LexerDiagnostic[] = [];

    constructor(source: string) {
        this.source = source;
    }

    tokenize(): LexResult {
        while (!this.isAtEnd()) {
            this.scanToken();
        }
        this.addToken(TokenType.EOF, '', this.pos, 0);
        return { tokens: this.tokens, diagnostics: this.diagnostics };
    }

    // ─── Core scanning ─────────────────────────────────────────────

    private scanToken(): void {
        this.skipWhitespace();
        if (this.isAtEnd()) return;

        const c = this.current();

        // Newline
        if (c === '\n') {
            const start = this.pos;
            this.addToken(TokenType.Newline, '\n', start, 1);
            this.advance();
            this.line++;
            this.col = 0;
            return;
        }

        // Comments
        if (c === '/' && this.pos + 1 < this.source.length) {
            if (this.peek() === '/') {
                this.skipLineComment();
                return;
            }
            if (this.peek() === '*') {
                this.skipBlockComment();
                return;
            }
        }

        // String literal
        if (c === '"') {
            this.scanString();
            return;
        }

        // Char literal
        if (c === "'") {
            this.scanChar();
            return;
        }

        // Number literal
        if (this.isDigit(c)) {
            this.scanNumber();
            return;
        }

        // Identifier or keyword
        if (this.isAlpha(c)) {
            this.scanIdentifierOrKeyword();
            return;
        }

        // Operators and punctuation
        this.scanOperatorOrPunct();
    }

    // ─── Whitespace & Comments ──────────────────────────────────────

    private skipWhitespace(): void {
        while (!this.isAtEnd()) {
            const c = this.current();
            if (c === ' ' || c === '\t' || c === '\r') {
                this.advance();
            } else {
                break;
            }
        }
    }

    private skipLineComment(): void {
        // Skip //
        this.advance();
        this.advance();
        while (!this.isAtEnd() && this.current() !== '\n') {
            this.advance();
        }
        // Don't consume the newline — it will be emitted as TOKEN_NEWLINE
    }

    private skipBlockComment(): void {
        const startLine = this.line;
        const startCol = this.col;
        // Skip /*
        this.advance();
        this.advance();

        while (!this.isAtEnd()) {
            if (this.current() === '*' && this.pos + 1 < this.source.length && this.peek() === '/') {
                this.advance(); // *
                this.advance(); // /
                return;
            }
            if (this.current() === '\n') {
                this.line++;
                this.col = 0;
                this.pos++;
                continue;
            }
            this.advance();
        }

        // Unterminated
        this.diagnostics.push({
            line: startLine,
            col: startCol,
            endCol: startCol + 2,
            message: 'Unterminated block comment.',
            severity: 'error',
        });
    }

    // ─── String Literal ─────────────────────────────────────────────

    private scanString(): void {
        const startPos = this.pos;
        const startCol = this.col;
        const startLine = this.line;
        this.advance(); // skip opening "

        let value = '';

        while (!this.isAtEnd() && this.current() !== '"') {
            if (this.current() === '\n') {
                this.diagnostics.push({
                    line: startLine,
                    col: startCol,
                    endCol: this.col,
                    message: 'Unterminated string literal.',
                    severity: 'error',
                });
                // Don't consume the newline
                const lexeme = this.source.substring(startPos, this.pos);
                this.addToken(TokenType.StringLiteral, lexeme, startPos, this.pos - startPos);
                return;
            }
            if (this.current() === '\\') {
                this.advance(); // skip backslash
                if (this.isAtEnd()) break;
                const esc = this.current();
                if (VALID_ESCAPE_CHARS.has(esc)) {
                    switch (esc) {
                        case 'n': value += '\n'; break;
                        case 't': value += '\t'; break;
                        case '\\': value += '\\'; break;
                        case '"': value += '"'; break;
                        case '0': value += '\0'; break;
                        default: value += esc;
                    }
                } else {
                    this.diagnostics.push({
                        line: this.line,
                        col: this.col - 1,
                        endCol: this.col + 1,
                        message: `Unknown escape sequence: '\\${esc}'.`,
                        severity: 'warning',
                    });
                    value += esc;
                }
                this.advance();
            } else {
                value += this.current();
                this.advance();
            }
        }

        if (this.isAtEnd()) {
            this.diagnostics.push({
                line: startLine,
                col: startCol,
                endCol: this.col,
                message: 'Unterminated string literal.',
                severity: 'error',
            });
        } else {
            this.advance(); // skip closing "
        }

        const lexeme = this.source.substring(startPos, this.pos);
        this.addToken(TokenType.StringLiteral, lexeme, startPos, this.pos - startPos);
    }

    // ─── Char Literal ───────────────────────────────────────────────

    private scanChar(): void {
        const startPos = this.pos;
        const startCol = this.col;
        const startLine = this.line;
        this.advance(); // skip opening '

        if (this.isAtEnd() || this.current() === '\n') {
            this.diagnostics.push({
                line: startLine,
                col: startCol,
                endCol: this.col,
                message: 'Unterminated character literal.',
                severity: 'error',
            });
            const lexeme = this.source.substring(startPos, this.pos);
            this.addToken(TokenType.CharLiteral, lexeme, startPos, this.pos - startPos);
            return;
        }

        if (this.current() === '\\') {
            this.advance(); // skip backslash
            if (!this.isAtEnd()) {
                const esc = this.current();
                if (!VALID_ESCAPE_CHARS.has(esc)) {
                    this.diagnostics.push({
                        line: this.line,
                        col: this.col - 1,
                        endCol: this.col + 1,
                        message: `Unknown escape sequence: '\\${esc}'.`,
                        severity: 'warning',
                    });
                }
                this.advance();
            }
        } else {
            this.advance(); // skip the character
        }

        if (this.isAtEnd() || this.current() !== "'") {
            this.diagnostics.push({
                line: startLine,
                col: startCol,
                endCol: this.col,
                message: 'Unterminated character literal.',
                severity: 'error',
            });
        } else {
            this.advance(); // skip closing '
        }

        const lexeme = this.source.substring(startPos, this.pos);
        this.addToken(TokenType.CharLiteral, lexeme, startPos, this.pos - startPos);
    }

    // ─── Number Literal ─────────────────────────────────────────────

    private scanNumber(): void {
        const startPos = this.pos;
        let isFloat = false;

        while (!this.isAtEnd() && this.isDigit(this.current())) {
            this.advance();
        }

        // Check for decimal point (but NOT range operator ..)
        if (!this.isAtEnd() && this.current() === '.' &&
            this.pos + 1 < this.source.length && this.peek() !== '.') {
            if (this.isDigit(this.peek())) {
                isFloat = true;
                this.advance(); // skip .
                while (!this.isAtEnd() && this.isDigit(this.current())) {
                    this.advance();
                }
            }
        }

        const lexeme = this.source.substring(startPos, this.pos);
        this.addToken(
            isFloat ? TokenType.FloatLiteral : TokenType.IntLiteral,
            lexeme,
            startPos,
            this.pos - startPos
        );
    }

    // ─── Identifier / Keyword ───────────────────────────────────────

    private scanIdentifierOrKeyword(): void {
        const startPos = this.pos;

        while (!this.isAtEnd() && this.isAlphaNumeric(this.current())) {
            this.advance();
        }

        const lexeme = this.source.substring(startPos, this.pos);
        const keywordType = KEYWORD_MAP.get(lexeme);
        this.addToken(
            keywordType !== undefined ? keywordType : TokenType.Identifier,
            lexeme,
            startPos,
            this.pos - startPos
        );
    }

    // ─── Operators & Punctuation ────────────────────────────────────

    private scanOperatorOrPunct(): void {
        const startPos = this.pos;
        const c = this.current();
        this.advance();

        switch (c) {
            case '+':
                if (!this.isAtEnd() && this.current() === '=') {
                    this.advance();
                    this.addToken(TokenType.PlusAssign, '+=', startPos, 2);
                } else {
                    this.addToken(TokenType.Plus, '+', startPos, 1);
                }
                break;

            case '-':
                if (!this.isAtEnd() && this.current() === '>') {
                    this.advance();
                    this.addToken(TokenType.Arrow, '->', startPos, 2);
                } else if (!this.isAtEnd() && this.current() === '=') {
                    this.advance();
                    this.addToken(TokenType.MinusAssign, '-=', startPos, 2);
                } else {
                    this.addToken(TokenType.Minus, '-', startPos, 1);
                }
                break;

            case '*':
                if (!this.isAtEnd() && this.current() === '=') {
                    this.advance();
                    this.addToken(TokenType.StarAssign, '*=', startPos, 2);
                } else {
                    this.addToken(TokenType.Star, '*', startPos, 1);
                }
                break;

            case '/':
                if (!this.isAtEnd() && this.current() === '=') {
                    this.advance();
                    this.addToken(TokenType.SlashAssign, '/=', startPos, 2);
                } else {
                    this.addToken(TokenType.Slash, '/', startPos, 1);
                }
                break;

            case '%':
                this.addToken(TokenType.Percent, '%', startPos, 1);
                break;

            case '=':
                if (!this.isAtEnd() && this.current() === '=') {
                    this.advance();
                    this.addToken(TokenType.Eq, '==', startPos, 2);
                } else {
                    this.addToken(TokenType.Assign, '=', startPos, 1);
                }
                break;

            case '!':
                if (!this.isAtEnd() && this.current() === '=') {
                    this.advance();
                    this.addToken(TokenType.Neq, '!=', startPos, 2);
                } else {
                    this.addToken(TokenType.Not, '!', startPos, 1);
                }
                break;

            case '<':
                if (!this.isAtEnd() && this.current() === '=') {
                    this.advance();
                    this.addToken(TokenType.Lte, '<=', startPos, 2);
                } else {
                    this.addToken(TokenType.Lt, '<', startPos, 1);
                }
                break;

            case '>':
                if (!this.isAtEnd() && this.current() === '=') {
                    this.advance();
                    this.addToken(TokenType.Gte, '>=', startPos, 2);
                } else {
                    this.addToken(TokenType.Gt, '>', startPos, 1);
                }
                break;

            case '&':
                if (!this.isAtEnd() && this.current() === '&') {
                    this.advance();
                    this.addToken(TokenType.And, '&&', startPos, 2);
                } else {
                    this.diagnostics.push({
                        line: this.line,
                        col: this.col - 1,
                        endCol: this.col,
                        message: "Unexpected '&'. Use '@' for address-of, '&&' for logical AND.",
                        severity: 'error',
                    });
                    this.addToken(TokenType.Unknown, '&', startPos, 1);
                }
                break;

            case '|':
                if (!this.isAtEnd() && this.current() === '|') {
                    this.advance();
                    this.addToken(TokenType.Or, '||', startPos, 2);
                } else {
                    this.diagnostics.push({
                        line: this.line,
                        col: this.col - 1,
                        endCol: this.col,
                        message: "Unexpected '|'. Use '||' for logical OR.",
                        severity: 'error',
                    });
                    this.addToken(TokenType.Unknown, '|', startPos, 1);
                }
                break;

            case '@':
                this.addToken(TokenType.At, '@', startPos, 1);
                break;

            case '^':
                this.addToken(TokenType.Deref, '^', startPos, 1);
                break;

            case '{':
                this.addToken(TokenType.LBrace, '{', startPos, 1);
                break;

            case '}':
                this.addToken(TokenType.RBrace, '}', startPos, 1);
                break;

            case '(':
                this.addToken(TokenType.LParen, '(', startPos, 1);
                break;

            case ')':
                this.addToken(TokenType.RParen, ')', startPos, 1);
                break;

            case '[':
                this.addToken(TokenType.LBracket, '[', startPos, 1);
                break;

            case ']':
                this.addToken(TokenType.RBracket, ']', startPos, 1);
                break;

            case ';':
                this.addToken(TokenType.Semicolon, ';', startPos, 1);
                break;

            case ':':
                this.addToken(TokenType.Colon, ':', startPos, 1);
                break;

            case ',':
                this.addToken(TokenType.Comma, ',', startPos, 1);
                break;

            case '.':
                if (!this.isAtEnd() && this.current() === '.') {
                    this.advance();
                    this.addToken(TokenType.DotDot, '..', startPos, 2);
                } else {
                    this.addToken(TokenType.Dot, '.', startPos, 1);
                }
                break;

            default:
                this.diagnostics.push({
                    line: this.line,
                    col: this.col - 1,
                    endCol: this.col,
                    message: `Unexpected character: '${c}'.`,
                    severity: 'error',
                });
                this.addToken(TokenType.Unknown, c, startPos, 1);
                break;
        }
    }

    // ─── Helpers ────────────────────────────────────────────────────

    private current(): string {
        return this.source[this.pos];
    }

    private peek(): string {
        return this.source[this.pos + 1];
    }

    private advance(): void {
        this.pos++;
        this.col++;
    }

    private isAtEnd(): boolean {
        return this.pos >= this.source.length;
    }

    private isDigit(c: string): boolean {
        return c >= '0' && c <= '9';
    }

    private isAlpha(c: string): boolean {
        return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c === '_';
    }

    private isAlphaNumeric(c: string): boolean {
        return this.isAlpha(c) || this.isDigit(c);
    }

    private addToken(type: TokenType, lexeme: string, offset: number, length: number): void {
        this.tokens.push({
            type,
            lexeme,
            line: type === TokenType.Newline ? this.line : this.line,
            col: this.col - length,
            offset,
            length,
        });
    }
}

File: src/compiler/Lexer.cpp
#include "Lexer.h"
#include <cctype>

Lexer::Lexer(const std::string& source, const std::string& filename,
             ErrorReporter& errorReporter)
    : source_(source), filename_(filename), errors_(errorReporter),
      pos_(0), line_(1), column_(1) {
    initKeywords();
}

void Lexer::initKeywords() {
    // Type keywords
    keywords_["int"]         = TokenType::KW_INT;
    keywords_["float"]       = TokenType::KW_FLOAT;
    keywords_["double"]      = TokenType::KW_DOUBLE;
    keywords_["char"]        = TokenType::KW_CHAR;
    keywords_["string"]      = TokenType::KW_STRING;
    keywords_["bool"]        = TokenType::KW_BOOL;
    keywords_["void"]        = TokenType::KW_VOID;

    // Declaration keywords
    keywords_["declare"]     = TokenType::KW_DECLARE;
    keywords_["elite"]       = TokenType::KW_ELITE;
    keywords_["law"]         = TokenType::KW_LAW;
    keywords_["command"]     = TokenType::KW_COMMAND;

    // Entry point
    keywords_["regime"]      = TokenType::KW_REGIME;

    // Boolean literals
    keywords_["loyal"]       = TokenType::KW_LOYAL;
    keywords_["traitor"]     = TokenType::KW_TRAITOR;

    // Control flow
    keywords_["interrogate"] = TokenType::KW_INTERROGATE;
    keywords_["otherwise"]   = TokenType::KW_OTHERWISE;
    keywords_["impose"]      = TokenType::KW_IMPOSE;
    keywords_["from"]        = TokenType::KW_FROM;
    keywords_["forever"]     = TokenType::KW_FOREVER;
    keywords_["silence"]     = TokenType::KW_SILENCE;
    keywords_["proceed"]     = TokenType::KW_PROCEED;
    keywords_["report"]      = TokenType::KW_REPORT;

    // I/O
    keywords_["broadcast"]   = TokenType::KW_BROADCAST;
    keywords_["demand"]      = TokenType::KW_DEMAND;

    // Memory
    keywords_["summon"]      = TokenType::KW_SUMMON;
    keywords_["kill"]        = TokenType::KW_KILL;
    keywords_["null"]        = TokenType::KW_NULL_VAL;

    // Imports
    keywords_["import"]      = TokenType::KW_ENLIST;

    // Hashmaps
    keywords_["map"]         = TokenType::KW_MAP;

    // Smart Pointers (C++11) — Feature 1
    keywords_["guard"]        = TokenType::KW_GUARD;
    keywords_["share"]        = TokenType::KW_SHARE;
    keywords_["summon_guard"] = TokenType::KW_SUMMON_GUARD;
    keywords_["summon_share"] = TokenType::KW_SUMMON_SHARE;

    // Lambda Expressions (C++11) — Feature 2
    keywords_["block"]       = TokenType::KW_BLOCK;

    // Auto Type Deduction (C++11) — Feature 3
    keywords_["auto"]        = TokenType::KW_AUTO;

    // Concurrency (C++11) — Feature 4
    keywords_["spawn"]       = TokenType::KW_SPAWN;

    // Structured Bindings (C++17) — Feature 5
    keywords_["unpack"]      = TokenType::KW_UNPACK;
}

char Lexer::current() const {
    if (isAtEnd()) return '\0';
    return source_[pos_];
}

char Lexer::peek() const {
    return peekAt(1);
}

char Lexer::peekAt(int offset) const {
    int idx = pos_ + offset;
    if (idx >= (int)source_.size()) return '\0';
    return source_[idx];
}

bool Lexer::isAtEnd() const {
    return pos_ >= (int)source_.size();
}

char Lexer::advance() {
    char c = source_[pos_];
    pos_++;
    column_++;
    return c;
}

void Lexer::skipWhitespaceAndComments() {
    while (!isAtEnd()) {
        char c = current();
        if (c == ' ' || c == '\t' || c == '\r') {
            advance();
        } else if (c == '/' && peek() == '/') {
            skipLineComment();
        } else if (c == '/' && peek() == '*') {
            if (!skipBlockComment()) return;
        } else {
            break;
        }
    }
}

void Lexer::skipLineComment() {
    // Skip until end of line (don't consume the newline itself).
    while (!isAtEnd() && current() != '\n') {
        advance();
    }
}

bool Lexer::skipBlockComment() {
    int startLine = line_;
    int startCol = column_;
    advance(); // skip /
    advance(); // skip *
    while (!isAtEnd()) {
        if (current() == '*' && peek() == '/') {
            advance(); // skip *
            advance(); // skip /
            return true;
        }
        if (current() == '\n') {
            line_++;
            column_ = 0; // advance() will set it to 1
        }
        advance();
    }
    errors_.error(filename_, startLine, startCol,
                  "Unterminated block comment.");
    return false;
}

std::vector<Token> Lexer::tokenize() {
    while (!isAtEnd()) {
        skipWhitespaceAndComments();
        if (isAtEnd()) break;
        scanToken();
    }
    addToken(TokenType::TOKEN_EOF, "", line_, column_);
    return tokens_;
}

void Lexer::scanToken() {
    char c = current();

    if (c == '\n') {
        addToken(TokenType::TOKEN_NEWLINE, "\\n", line_, column_);
        advance();
        line_++;
        column_ = 1;
        return;
    }

    if (c == '"') {
        scanString();
        return;
    }

    if (c == '\'') {
        scanChar();
        return;
    }

    if (std::isdigit(c)) {
        scanNumber();
        return;
    }

    if (std::isalpha(c) || c == '_') {
        scanIdentifierOrKeyword();
        return;
    }

    scanOperatorOrPunct();
}

void Lexer::scanString() {
    int startLine = line_;
    int startCol = column_;
    advance(); // skip opening "
    std::string value;
    while (!isAtEnd() && current() != '"') {
        if (current() == '\n') {
            errors_.error(filename_, startLine, startCol,
                          "Unterminated string literal.");
            return;
        }
        if (current() == '\\') {
            advance();
            switch (current()) {
                case 'n':  value += '\n'; break;
                case 't':  value += '\t'; break;
                case '\\': value += '\\'; break;
                case '"':  value += '"';  break;
                case '0':  value += '\0'; break;
                default:
                    errors_.warning(filename_, line_, column_,
                                    std::string("Unknown escape sequence: \\") + current());
                    value += current();
                    break;
            }
            advance();
        } else {
            value += current();
            advance();
        }
    }
    if (isAtEnd()) {
        errors_.error(filename_, startLine, startCol,
                      "Unterminated string literal.");
        return;
    }
    advance(); // skip closing "
    addToken(TokenType::TOKEN_STRING_LITERAL, value, startLine, startCol);
}

void Lexer::scanChar() {
    int startLine = line_;
    int startCol = column_;
    advance(); // skip opening '
    std::string value;
    if (isAtEnd() || current() == '\n') {
        errors_.error(filename_, startLine, startCol,
                      "Unterminated character literal.");
        return;
    }
    if (current() == '\\') {
        advance();
        switch (current()) {
            case 'n':  value += '\n'; break;
            case 't':  value += '\t'; break;
            case '\\': value += '\\'; break;
            case '\'': value += '\''; break;
            case '0':  value += '\0'; break;
            default:
                errors_.warning(filename_, line_, column_,
                                std::string("Unknown escape sequence: \\") + current());
                value += current();
                break;
        }
        advance();
    } else {
        value += current();
        advance();
    }
    if (isAtEnd() || current() != '\'') {
        errors_.error(filename_, startLine, startCol,
                      "Unterminated character literal.");
        return;
    }
    advance(); // skip closing '
    addToken(TokenType::TOKEN_CHAR_LITERAL, value, startLine, startCol);
}

void Lexer::scanNumber() {
    int startLine = line_;
    int startCol = column_;
    std::string num;
    bool isFloat = false;

    while (!isAtEnd() && std::isdigit(current())) {
        num += current();
        advance();
    }

    // Check for decimal point (but not .. range operator).
    if (!isAtEnd() && current() == '.' && peek() != '.') {
        isFloat = true;
        num += current();
        advance();
        while (!isAtEnd() && std::isdigit(current())) {
            num += current();
            advance();
        }
    }

    addToken(isFloat ? TokenType::TOKEN_FLOAT_LITERAL : TokenType::TOKEN_INT_LITERAL,
             num, startLine, startCol);
}

void Lexer::scanIdentifierOrKeyword() {
    int startLine = line_;
    int startCol = column_;
    std::string ident;

    while (!isAtEnd() && (std::isalnum(current()) || current() == '_')) {
        ident += current();
        advance();
    }

    auto it = keywords_.find(ident);
    if (it != keywords_.end()) {
        addToken(it->second, ident, startLine, startCol);
    } else {
        addToken(TokenType::TOKEN_IDENTIFIER, ident, startLine, startCol);
    }
}

void Lexer::scanOperatorOrPunct() {
    int startLine = line_;
    int startCol = column_;
    char c = current();

    switch (c) {
        case '+':
            advance();
            if (!isAtEnd() && current() == '=') {
                advance();
                addToken(TokenType::OP_PLUS_ASSIGN, "+=", startLine, startCol);
            } else {
                addToken(TokenType::OP_PLUS, "+", startLine, startCol);
            }
            break;

        case '-':
            advance();
            if (!isAtEnd() && current() == '>') {
                advance();
                addToken(TokenType::OP_ARROW, "->", startLine, startCol);
            } else if (!isAtEnd() && current() == '=') {
                advance();
                addToken(TokenType::OP_MINUS_ASSIGN, "-=", startLine, startCol);
            } else {
                addToken(TokenType::OP_MINUS, "-", startLine, startCol);
            }
            break;

        case '*':
            advance();
            if (!isAtEnd() && current() == '=') {
                advance();
                addToken(TokenType::OP_STAR_ASSIGN, "*=", startLine, startCol);
            } else {
                addToken(TokenType::OP_STAR, "*", startLine, startCol);
            }
            break;

        case '/':
            advance();
            if (!isAtEnd() && current() == '=') {
                advance();
                addToken(TokenType::OP_SLASH_ASSIGN, "/=", startLine, startCol);
            } else {
                addToken(TokenType::OP_SLASH, "/", startLine, startCol);
            }
            break;

        case '%':
            advance();
            addToken(TokenType::OP_PERCENT, "%", startLine, startCol);
            break;

        case '=':
            advance();
            if (!isAtEnd() && current() == '=') {
                advance();
                addToken(TokenType::OP_EQ, "==", startLine, startCol);
            } else {
                addToken(TokenType::OP_ASSIGN, "=", startLine, startCol);
            }
            break;

        case '!':
            advance();
            if (!isAtEnd() && current() == '=') {
                advance();
                addToken(TokenType::OP_NEQ, "!=", startLine, startCol);
            } else {
                addToken(TokenType::OP_NOT, "!", startLine, startCol);
            }
            break;

        case '<':
            advance();
            if (!isAtEnd() && current() == '=') {
                advance();
                addToken(TokenType::OP_LTE, "<=", startLine, startCol);
            } else {
                addToken(TokenType::OP_LT, "<", startLine, startCol);
            }
            break;

        case '>':
            advance();
            if (!isAtEnd() && current() == '=') {
                advance();
                addToken(TokenType::OP_GTE, ">=", startLine, startCol);
            } else {
                addToken(TokenType::OP_GT, ">", startLine, startCol);
            }
            break;

        case '&':
            advance();
            if (!isAtEnd() && current() == '&') {
                advance();
                addToken(TokenType::OP_AND, "&&", startLine, startCol);
            } else {
                errors_.error(filename_, startLine, startCol,
                              "Unexpected '&'. Use '@' for address-of, '&&' for logical AND.");
            }
            break;

        case '|':
            advance();
            if (!isAtEnd() && current() == '|') {
                advance();
                addToken(TokenType::OP_OR, "||", startLine, startCol);
            } else {
                errors_.error(filename_, startLine, startCol,
                              "Unexpected '|'. Use '||' for logical OR.");
            }
            break;

        case '@':
            advance();
            addToken(TokenType::OP_AT, "@", startLine, startCol);
            break;

        case '^':
            advance();
            addToken(TokenType::OP_DEREF, "^", startLine, startCol);
            break;

        case '{':
            advance();
            addToken(TokenType::PUNCT_LBRACE, "{", startLine, startCol);
            break;

        case '}':
            advance();
            addToken(TokenType::PUNCT_RBRACE, "}", startLine, startCol);
            break;

        case '(':
            advance();
            addToken(TokenType::PUNCT_LPAREN, "(", startLine, startCol);
            break;

        case ')':
            advance();
            addToken(TokenType::PUNCT_RPAREN, ")", startLine, startCol);
            break;

        case '[':
            advance();
            addToken(TokenType::PUNCT_LBRACKET, "[", startLine, startCol);
            break;

        case ']':
            advance();
            addToken(TokenType::PUNCT_RBRACKET, "]", startLine, startCol);
            break;

        case ';':
            advance();
            addToken(TokenType::PUNCT_SEMICOLON, ";", startLine, startCol);
            break;

        case ':':
            advance();
            addToken(TokenType::PUNCT_COLON, ":", startLine, startCol);
            break;

        case ',':
            advance();
            addToken(TokenType::PUNCT_COMMA, ",", startLine, startCol);
            break;

        case '.':
            advance();
            if (!isAtEnd() && current() == '.') {
                advance();
                addToken(TokenType::PUNCT_DOTDOT, "..", startLine, startCol);
            } else {
                addToken(TokenType::PUNCT_DOT, ".", startLine, startCol);
            }
            break;

        default:
            errors_.error(filename_, line_, column_,
                          std::string("Unexpected character: '") + c + "'.");
            advance();
            break;
    }
}

void Lexer::addToken(TokenType type, const std::string& lexeme, int startLine, int startCol) {
    tokens_.push_back(Token(type, lexeme, startLine, startCol));
}
